modules/runner/src/main/java/org/apache/ignite/internal/configuration/storage/LocalFileConfigurationStorage.java

```diff
@@ -0,0 +1,224 @@
+    private static final IgniteLogger LOG = Loggers.forClass(LocalFileConfigurationStorage.class);
+
+    private final Path configPath;
+    private final Map<String, Serializable> lastState = new HashMap<>();
+
+    private final AtomicReference<ConfigurationStorageListener> lsnrRef = new AtomicReference<>();
+
+    private final ExecutorService threadPool = Executors.newFixedThreadPool(4, new NamedThreadFactory("loc-cfg-file", LOG));
+
+    private final InFlightFutures futureTracker = new InFlightFutures();
+
+    public LocalFileConfigurationStorage(NodeBootstrapConfiguration configuration) {
+        this.configPath = configuration.config();
+        ensureConfigFile();
+    }
+
+    @Override
+    public CompletableFuture<Data> readDataOnRecovery() {
+        return readAllLatest("").thenApply(stringMap -> new Data(stringMap, 0));
+    }
+
+    @Override
+    public CompletableFuture<Map<String, ? extends Serializable>> readAllLatest(String prefix) {
+        ensureConfigFile();
+        Map<String, Serializable> map = new HashMap<>();
+        if (!lastState.isEmpty()) {
+            for (Entry<String, ConfigValue> e : parseConfigOptions().entrySet()) {
+                String key = e.getKey();
+                if (key.startsWith(prefix)) {
+                    ConfigValue value = e.getValue();
+                    if (value.valueType() == ConfigValueType.LIST) {
+                        map.put(key, toArray(lastState.get(key).getClass(), (ConfigList) value));
+                    } else {
+                        map.put(key, (Serializable) value.unwrapped());
+                    }
+                }
+            }
+        }
+        return CompletableFuture.completedFuture(map);
+    }
+
+    @Override
+    public CompletableFuture<Serializable> readLatest(String key) {
+        ensureConfigFile();
+        return CompletableFuture.completedFuture((Serializable) parseConfigOptions().getValue(key).unwrapped());
+    }
+
+    @Override
+    public CompletableFuture<Boolean> write(Map<String, ? extends Serializable> newValues, long ver) {
+        ensureConfigFile();
+        saveValues(newValues);
+        lastState.putAll(newValues);
+        runAsync(() -> lsnrRef.get().onEntriesChanged(new Data(newValues, 0)));
+        return CompletableFuture.completedFuture(true);
+    }
+
+    private void runAsync(Runnable runnable) {
+        CompletableFuture<Void> future = CompletableFuture.runAsync(runnable, threadPool);
+
+        futureTracker.registerFuture(future);
+    }
+
+    @Override
+    public void registerConfigurationListener(ConfigurationStorageListener lsnr) {
+        if (!lsnrRef.compareAndSet(null, lsnr)) {
+            LOG.debug("Configuration listener has already been set");
+        }
+    }
+
+    @Override
+    public ConfigurationType type() {
+        return ConfigurationType.LOCAL;
+    }
+
+    @Override
+    public CompletableFuture<Long> lastRevision() {
+        return CompletableFuture.completedFuture(0L);
```

Comment:
Wow, why? Class is clearly not completed.
Too bad that `org.apache.ignite.internal.configuration.storage.ConfigurationStorageTest` is not sophisticated enough, that's on us...